fix: make process.execve's args argument optional

mureinik opened the pull request:
Align the code with the documentation and similar methods used to execute os commands - the `args` argument should be optional, and if omitted, treated as an empty array (`[]`).

Fixes: https://github.com/nodejs/node/issues/58411

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

lpinca commented:
Can you add a test?

mureinik commented:
> Can you please add a test?

@lpinca I'm actually a bit lost as how we'd test this, and would appreciate a pointer.
The other tests under (in `test-process-execve*.js`) call `process.execve` with the same node.js executable, the first argument as the same test, and some other args/env, and then assert that the replaced process gets the correct args/env passed to it.

For a test without any args, the best I could think of is something down the lines of replacing the process with something known and just asserting the process was replaced:
```javascript
process.execve(
    SOMETHING_GOES_HERE
);

// If process.execve succeed, this should never be executed.
fail('process.execve failed');
```

but I don't have a good idea for an executable that is :
1. Known to be available on any platform the CI runs on
2. Won't block once executed

The only two things I can think of that are known to be available is the node executable itself (`process.execPath`), but it opens an interactive repl and never returns and maybe the current shell (`process.env.SHELL`), but that would also open an interactive shell and never return.

If you can help point me at the right direction, I'd be happy to add a test.

lpinca commented:
> FWIW, on my local env I tested this with `process.execve('/bin/ls')`, but I don't think that in the CI we can make assumptions on the location of `ls`.

What do think about finding the location first? Something like this

```js
const path = child_process.spawnSync('which', ['ls'], { encoding: 'utf8' }).stdout.trim()
```

mureinik commented:
@lpinca I was trying to avoid building a test that relies on `child_process`, but I guess these are more like integration tests than unit tests - the existing tests rely on `worker_threads` and (indirectly) other `process` properties/functions anyway, so this is probably a moot point.

I've added such a test, and rebased on top of the latest `main`.

For the record - I'm using `ps` instead of `ls`. On some platforms (e.g., my laptop), calling `ls` without arguments spews a usage message to `stderr`, which causes `test.py` to fail the test.

mureinik commented:
This is embarrassing - my laptop is a mac, and the new test passes locally on it. On the CI env, macOS is the only thing that fails :-(
I'm looking in to it.

mureinik commented:
Using `ps` failed the macOS CI. I'm trying with `date` to help triage what's going on here - i.e., is this a problem with `ps`, or something fundamentally different in the CI compared to my local env.

mureinik commented:
@lpinca I'm not sure what exactly went wrong with using `ps` there, but TBH, I'd rather not dig too deeply into it.
Using `date` did the trick, and it seems common enough in *nix-like systems that it can be used for these tests.

lpinca commented on test/parallel/test-process-execve-no-args.js:
```diff
@@ -0,0 +1,24 @@
+const { isMainThread } = require('worker_threads');
+const { spawnSync } = require('child_process');
+
+if (!isMainThread) {
+  skip('process.execve is not available in Workers');
```
I see this is also used in other test files for `process.execve()` and it was suggested in https://github.com/nodejs/node/pull/56496#discussion_r1905035954, but I don't understand why it is needed? The test does not use worker threads nor `node:test`.

richardlau replied:
We run the entire test suite with `test.py --worker` in [node-test-commit-custom-suites-freestyle (test-worker)](https://ci.nodejs.org/job/node-test-commit-custom-suites-freestyle/).

lpinca replied:
Oh ok, thanks.

richardlau commented on test/parallel/test-process-execve-no-args.js:
```diff
@@ -0,0 +1,24 @@
+  skip('process.execve is not available in Windows or IBM i');
+}
+
+// Get full path to the ps executable
```
This comment no longer matches the code.

mureinik replied:
@richardlau fixed. I'm not sure what Node's etiquette for resolving PR comments, so to be on the safe side I'll leave the comment unresolved so you can resolve it if you think the comment is OK now.

mureinik commented:
@richardlau, apologies for nagging, but it looks like this PR lost its steam. I updated it to handle the (relevant) errors seen in the Jenkins CI, but IIUC, I can't re-trigger that run.
Can it please be retriggered so we can move forward with this?

mureinik commented:
@richardlau Thanks!

Looking at the failures, [node-test-linux-alpine-last-latest-x64](https://ci.nodejs.org/job/node-test-commit-linux/nodes=alpine-last-latest-x64/65075/) fails on the startup time of node (see https://ci.nodejs.org/job/node-test-commit-linux/nodes=alpine-last-latest-x64/65075/testReport/junit/(root)/sequential/test_perf_hooks/) that probably isn't related to this MR (i.e., starting up node shouldn't call `execve`).

In the other two failed jobs, it shows that there are no tests. IIUC, these are just aggregate jobs, and the "real" failure is the aforementioned startup time failure.

I'm not quite sure how to proceed from here.

mureinik commented:
Following up on my prev comment - for the record, now the CI passes. I'm assuming this means the tests that failed aren't 100% stable, and as explained above, are probably unrelated to the PR's content.

mureinik commented:
@bnoordhuis @lpinca @richardlau this kind of lost steam...
Is there anything else this PR needs?
